### PR TITLE
Implement read-only UI in HASH app

### PR DESCRIPTION
**Pull request by teenoh**
## 🌟 What is the purpose of this PR?

This PR includes the frontend work needed to handle read-only mode in the HASH app. The read-only version of pages can be accessed by attaching `?readonly` to the page's url.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1202621375435411/1202643260631375) _(internal)_

## 🔍 What does this change?

- Adds a `ReadonlyMode` context provider that determines if the page is in readonly mode. Right now we use the presence of `readonly` query parameter in the page's url to check this but a better approach will be implemented later.
-  Sets Prosemirror's editable prop to false when in read-only mode (this prevents editing in text blocks).
- Hides `Sidebar` and `BlockHandle` when in read-only mode.
- Updates `write-based` graph functions to throw an error when in read-only mode. `read-based` functions like `getEntity`, `aggregateEntities` will still work in this state
- Updates Code block to handle read-only mode
- Updates the readme file to include `NEXT_PUBLIC_BLOCKPROTOCOL_API_KEY`

## ⚠️ Known issues

- Since this only involves FE work, unauthorized users will not be able to access the page even if it is in readonly mode

## 🐾 Next steps

- Update existing blocks to implement read-only mode
- Add playwright tests for read-only mode

## 🛡 What tests cover this?

- Playwright tests for this state will be added in a follow up

## ❓ How to test this?

1.  Go to any page
1.  Attach `?readonly` to the page's url and refresh

## 📹 Demo

https://user-images.githubusercontent.com/16928688/182838688-3fa28b54-c90b-47fd-b1b5-07b9fcc3e459.mov

### CODE REVIEW

File: packages/hash/frontend/src/shared/readonly-mode/context.tsx [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+export const ReadonlyModeProvider = ({
+  children,
+}: {
+  children?: ReactNode;
+}) => {
+  const router = useRouter();
+
+  const contextValue = useMemo(
+    () => ({
+      readonlyMode: "readonly" in router.query,
```
Comment (kachkaev) [inline packages/hash/frontend/src/shared/readonly-mode/context.tsx:21]: We know that this is a temporary solution but this is not obvious from the codebase itself. If someone opens this file, they may get confused. I’d place a comment to communicate our intent to others and future selves. Perhaps, the right place for this is a JSDoc for `const ReadonlyModeProvider`

File: packages/hash/frontend/src/shared/readonly-mode/context.tsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import { useRouter } from "next/router";
+import { createContext, ReactNode, useContext, useMemo } from "react";
+
+type ReadonlyModeInfo = {
```
Comment (kachkaev) [inline packages/hash/frontend/src/shared/readonly-mode/context.tsx:4]:
```suggestion
export type ReadonlyModeInfo = {
```

File: packages/hash/frontend/src/components/RemoteBlock/RemoteBlock.tsx [hunk 3]
```diff
@@ -83,6 +84,14 @@ export const RemoteBlock: FunctionComponent<RemoteBlockProps> = ({
     }
   }, [graphProperties.linkedAggregations, graphService]);
 
+  useEffect(() => {
+    if (graphService) {
+      graphService.readonly({
+        data: graphProperties.readonly,
+      });
+    }
+  }, [graphProperties.readonly, graphService]);
```
Comment (kachkaev) [inline packages/hash/frontend/src/components/RemoteBlock/RemoteBlock.tsx:93]: @nathggns do you see any better ways of syncing context with graph service (and ProseMirror too)? If the is one, we can implement it in a follow-up.

**Review by kachkaev — CHANGES_REQUESTED** [review]
This looks very good @teenoh! I left a few minor suggestions to consider. I tried changing `readonly` on the fly and the app did not crash, which is good to know. Here is how I hacked the toggle locally inside `ReadonlyModeProvider`:

```ts
  const [readonlyMode, setReadonlyMode] = useState(true);

  useEffect(() => {
    window.document.body.addEventListener("keyup", ({ key }) => {
      if (key === "Escape") {
        setReadonlyMode((value) => !value);
      }
    });
  }, []);
```

The only strange thing I noticed was this: pressing on an callout icon in readonly mode was triggering a broken network request instead of a local exception. Can you please take a look?

`http://localhost:5001/graphql?updatePage`

```
{"errors":[{"message":"Variable \"$properties\" got invalid value { text: { __linkedData: [Object], data: [Object] }, icon: \"💡\" }; Field \"text\" is not defined by type \"PageUpdateData\".","extensions":{"code":"BAD_USER_INPUT","exception":{"stacktrace":["GraphQLError: Variable \"$properties\" got invalid value { text: { __linkedData: [Object], data: [Object] }, icon: \"💡\" }; Field \"text\" is not defined by type \"PageUpdateData\".","    at /path/to/hash/node_modules/graphql/execution/values.js:116:15","    at coerceInputValueImpl (/path/to/hash/node_modules/graphql/utilities/coerceInputValue.js:114:9)","    at coerceInputValueImpl (/path/to/hash/node_modules/graphql/utilities/coerceInputValue.js:54:14)","    at coerceInputValue (/path/to/hash/node_modules/graphql/utilities/coerceInputValue.js:37:10)","    at _loop (/path/to/hash/node_modules/graphql/execution/values.js:109:69)","    at coerceVariableValues (/path/to/hash/node_modules/graphql/execution/values.js:121:16)","    at getVariableValues (/path/to/hash/node_modules/graphql/execution/values.js:50:19)","    at buildExecutionContext (/path/to/hash/node_modules/graphql/execution/execute.js:203:61)","    at executeImpl (/path/to/hash/node_modules/graphql/execution/execute.js:101:20)","    at execute (/path/to/hash/node_modules/graphql/execution/execute.js:60:35)","    at execute (/path/to/hash/node_modules/apollo-server-core/src/requestPipeline.ts:500:34)","    at processGraphQLRequest (/path/to/hash/node_modules/apollo-server-core/src/requestPipeline.ts:406:28)","    at runMicrotasks (<anonymous>)","    at processTicksAndRejections (node:internal/process/task_queues:96:5)","    at processHTTPRequest (/path/to/hash/node_modules/apollo-server-core/src/runHttpQuery.ts:433:24)"]}}},{"message":"Variable \"$properties\" got invalid value { text: { __linkedData: [Object], data: [Object] }, icon: \"💡\" }; Field \"icon\" is not defined by type \"PageUpdateData\".","extensions":{"code":"BAD_USER_INPUT","exception":{"stacktrace":["GraphQLError: Variable \"$properties\" got invalid value { text: { __linkedData: [Object], data: [Object] }, icon: \"💡\" }; Field \"icon\" is not defined by type \"PageUpdateData\".","    at /path/to/hash/node_modules/graphql/execution/values.js:116:15","    at coerceInputValueImpl (/path/to/hash/node_modules/graphql/utilities/coerceInputValue.js:114:9)","    at coerceInputValueImpl (/path/to/hash/node_modules/graphql/utilities/coerceInputValue.js:54:14)","    at coerceInputValue (/path/to/hash/node_modules/graphql/utilities/coerceInputValue.js:37:10)","    at _loop (/path/to/hash/node_modules/graphql/execution/values.js:109:69)","    at coerceVariableValues (/path/to/hash/node_modules/graphql/execution/values.js:121:16)","    at getVariableValues (/path/to/hash/node_modules/graphql/execution/values.js:50:19)","    at buildExecutionContext (/path/to/hash/node_modules/graphql/execution/execute.js:203:61)","    at executeImpl (/path/to/hash/node_modules/graphql/execution/execute.js:101:20)","    at execute (/path/to/hash/node_modules/graphql/execution/execute.js:60:35)","    at execute (/path/to/hash/node_modules/apollo-server-core/src/requestPipeline.ts:500:34)","    at processGraphQLRequest (/path/to/hash/node_modules/apollo-server-core/src/requestPipeline.ts:406:28)","    at runMicrotasks (<anonymous>)","    at processTicksAndRejections (node:internal/process/task_queues:96:5)","    at processHTTPRequest (/path/to/hash/node_modules/apollo-server-core/src/runHttpQuery.ts:433:24)"]}}}]}
```

File: packages/hash/frontend/src/components/BlockLoader/BlockLoader.tsx [hunk 4]
```diff
@@ -78,19 +79,29 @@ export const BlockLoader: FunctionComponent<BlockLoaderProps> = ({
...
-  const { createEntity } = useBlockProtocolCreateEntity(accountId);
-  const { createEntityType } = useBlockProtocolCreateEntityType(accountId);
-  const { deleteLinkedAggregation } = useBlockProtocolDeleteLinkedAggregation();
-  const { deleteLink } = useBlockProtocolDeleteLink();
-  const { updateEntity } = useBlockProtocolUpdateEntity();
-  const { uploadFile } = useBlockProtocolFileUpload(accountId);
-  const { updateEntityType } = useBlockProtocolUpdateEntityType();
-  const { updateLinkedAggregation } = useBlockProtocolUpdateLinkedAggregation();
+  const { createLinkedAggregation } =
+    useBlockProtocolCreateLinkedAggregation(readonlyMode);
```
[the remark this answers is not in the document]
Comment (teenoh) [reply packages/hash/frontend/src/components/BlockLoader/BlockLoader.tsx:87]: Thanks. I initially thought about that but I didn't want to tightly couple the hooks to `ReadonlyMode` context. If we decide to move all the hooks to a different project where `ReadonlyMode` context provider isn't present, the hooks will break

File: packages/hash/frontend/src/components/RemoteBlock/RemoteBlock.tsx [hunk 5]
```diff
@@ -2,6 +2,7 @@ import { BlockMetadata, UnknownRecord } from "@blockprotocol/core";
 import {
   BlockGraphProperties,
   EmbedderGraphMessageCallbacks,
+  // useGraphEmbedderService
```
[the remark this answers is not in the document]
Comment (teenoh) [reply packages/hash/frontend/src/components/RemoteBlock/RemoteBlock.tsx:5]: Thanks for catching

File: packages/hash/frontend/src/pages/_app.page.tsx [hunk 6]
```diff
@@ -77,11 +78,13 @@ const App: FunctionComponent<AppProps> = ({
...
-                {getLayout(<Component {...pageProps} />)}
-              </RoutePageInfoProvider>
-            </RouteAccountInfoProvider>
+            <ReadonlyModeProvider>
+              <RouteAccountInfoProvider>
+                <RoutePageInfoProvider>
+                  {getLayout(<Component {...pageProps} />)}
+                </RoutePageInfoProvider>
+              </RouteAccountInfoProvider>
+            </ReadonlyModeProvider>
```
[the remark this answers is not in the document]
Comment (teenoh) [reply packages/hash/frontend/src/pages/_app.page.tsx:87]: I agree with you, will update

File: packages/hash/frontend/src/shared/readonly-mode/context.tsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  );
+
+  return (
+    <ReadonlyModeContext.Provider value={contextValue}>
+      {children}
+    </ReadonlyModeContext.Provider>
+  );
+};
+
+export const useReadonlyMode = () => {
```
[the remark this answers is not in the document]
Comment (teenoh) [reply packages/hash/frontend/src/shared/readonly-mode/context.tsx:33]: Thanks for reminding me about this. Will update!

File: packages/hash/frontend/src/blocks/page/BlockHandle.tsx [hunk 8]
```diff
@@ -66,8 +68,18 @@ const BlockHandle: ForwardRefRenderFunction<
...
+  if (readonlyMode) {
+    return null;
+  }
+
   return (
-    <Box ref={ref} data-testid="block-handle">
+    <Box
+      ref={ref}
+      sx={{
+        display: readonlyMode ? "block" : "block",
```
[the remark this answers is not in the document]
Comment (teenoh) [reply packages/hash/frontend/src/blocks/page/BlockHandle.tsx:79]: Thanks for catching!

File: packages/hash/frontend/src/blocks/page/BlockHandle.tsx [hunk 9]
```diff
@@ -11,6 +11,7 @@ import { BlockConfigMenu } from "./BlockConfigMenu/BlockConfigMenu";
 import { BlockContextMenu } from "./BlockContextMenu/BlockContextMenu";
 import { useBlockView } from "./BlockViewContext";
 import { useBlockContext } from "./BlockContext";
+import { useReadonlyMode } from "../../shared/readonly-mode/context";
```
[the remark this answers is not in the document]
Comment (teenoh) [reply packages/hash/frontend/src/blocks/page/BlockHandle.tsx:14]: Makes sense. Thanks!

**Comment by teenoh** [comment]
Thanks @kachkaev 

> The only strange thing I noticed was this: pressing on an callout icon in readonly mode was triggering a broken network request instead of a local exception. Can you please take a look?
> http://localhost:5001/graphql?updatePage

Thanks for pointing this out, will have a look to see what's wrong

File: packages/hash/frontend/src/components/RemoteBlock/RemoteBlock.tsx (hunk 3 above)
Comment (teenoh) [reply packages/hash/frontend/src/components/RemoteBlock/RemoteBlock.tsx:93]: What problems do you see with this approach @kachkaev ?

File: packages/hash/frontend/src/blocks/page/BlockHandle.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (teenoh) [reply packages/hash/frontend/src/blocks/page/BlockHandle.tsx]: Updated in [a74f0d5](https://github.com/hashintel/hash/pull/893/commits/a74f0d55f7c9b7806d3b5ab86f07c41f1644fe6a)

File: packages/hash/frontend/src/blocks/page/BlockHandle.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (teenoh) [reply packages/hash/frontend/src/blocks/page/BlockHandle.tsx]: Updated in [a74f0d5](https://github.com/hashintel/hash/pull/893/commits/a74f0d55f7c9b7806d3b5ab86f07c41f1644fe6a)

File: packages/hash/frontend/src/components/RemoteBlock/RemoteBlock.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (teenoh) [reply packages/hash/frontend/src/components/RemoteBlock/RemoteBlock.tsx]: Updated in [a74f0d5](https://github.com/hashintel/hash/pull/893/commits/a74f0d55f7c9b7806d3b5ab86f07c41f1644fe6a)

File: packages/hash/frontend/src/shared/readonly-mode/context.tsx [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+}: {
+  children?: ReactNode;
+}) => {
+  const router = useRouter();
+
+  const contextValue = useMemo(
+    () => ({
+      readonlyMode: "readonly" in router.query,
+    }),
+    [router],
```
[the remark this answers is not in the document]
Comment (teenoh) [reply packages/hash/frontend/src/shared/readonly-mode/context.tsx]: Addressed in [a74f0d5](https://github.com/hashintel/hash/pull/893/commits/a74f0d55f7c9b7806d3b5ab86f07c41f1644fe6a)

File: packages/hash/frontend/src/shared/readonly-mode/context.tsx (hunk 1 above)
Comment (teenoh) [reply packages/hash/frontend/src/shared/readonly-mode/context.tsx]: I agree. Updated in [a74f0d5](https://github.com/hashintel/hash/pull/893/commits/a74f0d55f7c9b7806d3b5ab86f07c41f1644fe6a)

File: packages/hash/frontend/src/shared/readonly-mode/context.tsx (hunk 2 above)
Comment (teenoh) [reply packages/hash/frontend/src/shared/readonly-mode/context.tsx]: Updated in [a74f0d5](https://github.com/hashintel/hash/pull/893/commits/a74f0d55f7c9b7806d3b5ab86f07c41f1644fe6a)

File: packages/hash/frontend/src/components/RemoteBlock/RemoteBlock.tsx (hunk 3 above)
Comment (kachkaev) [reply packages/hash/frontend/src/components/RemoteBlock/RemoteBlock.tsx:92]: `useEffect` creates a small time gap between a change in `graphProperties.readonly` and `graphService` state. This _might_ cause an ‘impossible’ app state and cause a crash. I don’t know enough to be sure one way or the other, so curious what others think. We can leave this as is for now I guess.

**Review by kachkaev — APPROVED** [review]
Thanks for addressing PR comments @teenoh. The diff looks great!

**Pull request merged**